File: tests/mg_checks_regression.f90
program mg_checks_regression
  use mg_checks, only: perform_check
  use handling_processes, only: pgl,ngroups
  use run_parameters, only: reset_run_parameters,alphaEW
  implicit none
  character(len=32) :: mode

  mode='pass'
  if (command_argument_count().ge.1) call get_command_argument(1,mode)
  call reset_run_parameters()
  ngroups=1
  allocate(pgl(1))
  pgl(1)%next=4
  pgl(1)%nproc=1
  allocate(pgl(1)%amp2(1),pgl(1)%iden(1),pgl(1)%amps(1))
  allocate(pgl(1)%amps(1)%n_sing(1))
  pgl(1)%amp2=1d0
  pgl(1)%iden=1_8
  pgl(1)%amps(1)%n_sing=0

  select case (trim(mode))
  case ('pass')
     call perform_check(1,1)
     write(*,'(a)') 'MadGraph comparison regression: PASS'
  case ('zero')
     pgl(1)%amp2=0d0
     call perform_check(1,1)
     write(*,'(a)') 'MadGraph zero comparison regression: PASS'
  case ('mismatch','bad-reference','bad-momentum','bad-count','truncated','trailing')
     call perform_check(1,1)
     error stop 'invalid MadGraph reference unexpectedly succeeded'
  case ('coupling-overflow')
     pgl(1)%amps(1)%n_sing=2
     alphaEW=huge(1d0)/100d0
     call perform_check(1,1)
     error stop 'overflowing coupling normalization unexpectedly succeeded'
  case ('overflow')
     pgl(1)%amp2=huge(1d0)
     call perform_check(1,1)
     error stop 'overflowing MadGraph normalization unexpectedly succeeded'
  case ('negative-code')
     pgl(1)%amp2=-1d0
     call perform_check(1,1)
     error stop 'negative matrix element unexpectedly succeeded'
  case ('bad-next')
     pgl(1)%next=21
     call perform_check(1,1)
     error stop 'invalid process dimension unexpectedly succeeded'
  case default
     error stop 'unknown MadGraph regression mode'
  end select
end program mg_checks_regression

File: tests/mg_checks_regression.py
#!/usr/bin/env python3
"""Validate MadGraph reference parsing and relative comparisons synthetically."""

from __future__ import annotations

import math
import pathlib
import subprocess
import sys
import tempfile


def run(executable: pathlib.Path, mode: str, cwd: pathlib.Path) -> subprocess.CompletedProcess[str]:
    return subprocess.run(
        [str(executable), mode],
        cwd=cwd,
        text=True,
        stdout=subprocess.PIPE,
        stderr=subprocess.STDOUT,
        check=False,
    )


def write_reference(path: pathlib.Path, count: int, values: list[str], momentum: str = "0 0 0 0") -> None:
    rows = [momentum] * 4 + [str(count)] + values
    path.write_text("\n".join(rows) + "\n", encoding="utf-8")


def main() -> int:
    executable = pathlib.Path(sys.argv[1]).resolve()
    normalization = (4.0 * math.pi * 0.118) ** 2
    with tempfile.TemporaryDirectory(prefix="amplicol_mg_") as directory:
        cwd = pathlib.Path(directory)
        reference_dir = cwd / "Utilities" / "ME_checks"
        reference_dir.mkdir(parents=True)
        reference = reference_dir / "momenta_1_1.txt"

        write_reference(reference, 2, ["0.0", f"{normalization * (1.0 + 5.0e-5):.17e}"])
        passed = run(executable, "pass", cwd)
        if passed.returncode != 0 or "comparison regression: PASS" not in passed.stdout:
            raise RuntimeError(f"valid MadGraph comparison failed\n{passed.stdout}")

        write_reference(reference, 1, ["0.0"])
        zero = run(executable, "zero", cwd)
        if zero.returncode != 0 or "zero comparison regression: PASS" not in zero.stdout:
            raise RuntimeError(f"zero MadGraph comparison failed\n{zero.stdout}")

        cases: list[tuple[str, str]] = []
        write_reference(reference, 1, [f"{normalization * 1.01:.17e}"])
        cases.append(("mismatch", "disagreement with MadGraph"))
        for mode, expected in cases:
            result = run(executable, mode, cwd)
            if result.returncode == 0 or expected not in result.stdout:
                raise RuntimeError(f"{mode} was not rejected\n{result.stdout}")

        write_reference(reference, 1, ["-1.0"])
        result = run(executable, "bad-reference", cwd)
        if result.returncode == 0 or "invalid matrix elements" not in result.stdout:
            raise RuntimeError(f"negative reference was not rejected\n{result.stdout}")

        write_reference(reference, 1, ["NaN"])
        result = run(executable, "bad-reference", cwd)
        if result.returncode == 0 or "invalid matrix elements" not in result.stdout:
            raise RuntimeError(f"non-finite reference was not rejected\n{result.stdout}")

        write_reference(reference, 1, ["1.0"], momentum="NaN 0 0 0")
        result = run(executable, "bad-momentum", cwd)
        if result.returncode == 0 or "non-finite momentum" not in result.stdout:
            raise RuntimeError(f"non-finite momentum was not rejected\n{result.stdout}")

        write_reference(reference, 1_000_001, [])
        result = run(executable, "bad-count", cwd)
        if result.returncode == 0 or "Invalid MadGraph reference count" not in result.stdout:
            raise RuntimeError(f"oversized reference count was not rejected\n{result.stdout}")

        write_reference(reference, 1, [])
        result = run(executable, "truncated", cwd)
        if result.returncode == 0 or "Could not read MadGraph reference value" not in result.stdout:
            raise RuntimeError(f"truncated reference was not rejected\n{result.stdout}")

        write_reference(reference, 1, ["1.0", "2.0"])
        result = run(executable, "trailing", cwd)
        if result.returncode == 0 or "contains trailing data" not in result.stdout:
            raise RuntimeError(f"trailing reference data was not rejected\n{result.stdout}")

        for mode, expected in {
            "overflow": "overflows MadGraph comparison normalization",
            "coupling-overflow": "normalization cannot be represented",
            "negative-code": "Invalid normalization metadata",
            "bad-next": "Invalid process dimensions",
        }.items():
            result = run(executable, mode, cwd)
            if result.returncode == 0 or expected not in result.stdout:
                raise RuntimeError(f"{mode} state was not rejected\n{result.stdout}")
    print("MadGraph malformed-reference regression: PASS")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
